internal/app/admin/http/handler.go

```diff
@@ -109,52 +109,67 @@ func configDumpHandler(bootstrapConfig *bootstrapv1.Bootstrap) http.HandlerFunc
 	}
 }
 
-// TODO(lisalu): Support dump of matching resources when cache key regex is provided.
+func printCacheEntries(keys []string, cache cache.ReadOnlyCache, w http.ResponseWriter) {
+	resp := marshallableCache{}
+	for _, key := range keys {
+		resource, err := cache.FetchReadOnly(key)
+		if err != nil {
+			w.WriteHeader(http.StatusNotFound)
+			fmt.Fprintf(w, "no resource for key %s found in cache.\n", key)
+			continue
+		}
+		resp.Cache = append(resp.Cache, resourceToPayload(key, resource)...)
+	}
+	resourceString, err := stringify.InterfaceToString(resp)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		fmt.Fprintf(w, "unable to convert resource to string.\n")
+		return
+	}
+
+	if len(resp.Cache) > 0 {
+		fmt.Fprintf(w, "%s\n", resourceString)
+	}
+
+}
+
 func cacheDumpHandler(o *orchestrator.Orchestrator) http.HandlerFunc {
 	return func(w http.ResponseWriter, req *http.Request) {
 		cacheKey := getParam(req.URL.Path)
 		c := orchestrator.Orchestrator.GetReadOnlyCache(*o)
-		resp := marshallableCache{}
+		var keysToPrint []string
+
+		// If wildcard suffix provided, output all cache entries that match given prefix.
 		// If no key is provided, output the entire cache.
-		if cacheKey == "" {
-			keys, err := orchestrator.Orchestrator.GetDownstreamAggregatedKeys(*o)
+		containsWildcardSuffix := strings.HasSuffix(cacheKey, "*")
+		if cacheKey == "" || containsWildcardSuffix {
```

Comment:
What do you think of making a separate function:
```
// hasWildcardSuffix returns whether the supplied key contains an empty string or a * suffix.
// Return true in these scenarios and the prefix.
func hasWildcardSuffix(key string) (bool, string) {
    return if key == "" || strings.HasSuffix(key, "*"), strings.TrimSuffix(key, "*")
}
```

I think this way it can help get rid of some of the nested ifs between the two pathways and treat them equally, just testing for the suffix.

Reply:
I think I'm misinterpreting your function here, but I don't understand how it removes the need of two pathways. The original goal of having two pathways was to just avoid doing prefix checks for the "print entire cache" case since it was unnecessary. I don't think it's an expensive operation, so I can go ahead and remove the nested ifs and treat is as one case. Is this what you mean?

```
               if cacheKey == "" || containsWildcardSuffix {
			// Retrieve all keys
			allKeys, err := orchestrator.Orchestrator.GetDownstreamAggregatedKeys(*o)
			if err != nil {
				w.WriteHeader(http.StatusInternalServerError)
				fmt.Fprintf(w, "error in getting cache keys: %s", err.Error())
				return
			}

			// Find keys that match prefix of wildcard
			rootCacheKeyName := strings.TrimSuffix(cacheKey, "*")
			for potentialMatchKey := range allKeys {
				if strings.HasPrefix(potentialMatchKey, rootCacheKeyName) {
					keysToPrint = append(keysToPrint, potentialMatchKey)
				}
			}
		} ....
```

Reply:
My version of a shorter function 
```
func cacheDumpHandler(o *orchestrator.Orchestrator) http.HandlerFunc {
	return func(w http.ResponseWriter, req *http.Request) {
		cacheKey := getParam(req.URL.Path)
		c := orchestrator.Orchestrator.GetReadOnlyCache(*o)
		var keysToPrint []string

		// If wildcard suffix provided, output all cache entries that match given prefix.
		// If no key is provided, output the entire cache.
               if cacheKey == "" {
                 cacheKey = "*"
               }
		containsWildcardSuffix := strings.HasSuffix(cacheKey, "*")
		if containsWildcardSuffix {
			// Retrieve all keys
			allKeys, err := orchestrator.Orchestrator.GetDownstreamAggregatedKeys(*o)
			if err != nil {
				w.WriteHeader(http.StatusInternalServerError)
				fmt.Fprintf(w, "error in getting cache keys: %s", err.Error())
				return
			}

                       rootCacheKeyName := strings.TrimSuffix(cacheKey, "*")
		       for potentialMatchKey := range allKeys {
			if strings.HasPrefix(potentialMatchKey, rootCacheKeyName) {
					keysToPrint = append(keysToPrint, potentialMatchKey)
				}
			}
		} else {
			// Otherwise return the cache entry corresponding to the given key.
			keysToPrint = []string{cacheKey}
		}
		printCacheEntries(keysToPrint, c, w)
	}
}
```

Reply:
@samrabelachew I just meant to get rid of [this](https://github.com/envoyproxy/xds-relay/pull/139/files#diff-29e3fee34912e53bdb13be1567ad0307R155-L133) nested block to improve readability, rather than perf gains. There's multiple ways to go about it including the one Jyoti listed, I'm ok with either.